E2E tests setup and initial tests

mnzaki opened the pull request:
Resolves #1415 

~This builds on #1410 which must be merged first~

## NOTES
- Android `minSdkVersion` bumped to 18
- iOS not tested and may require more setup
- No release build testing yet
- `yarn e2e:android` to run the tests
- `yarn e2e:android -r` to _reuse_ last app installation and prevent detox from reinstalling
- `yarn e2e:android -r recovery` to run the tests matching "recovery" (this is somewhat broken)

VolkerSchiewe commented on android/app/build.gradle:
```diff
@@ -111,6 +111,8 @@ android {
         vectorDrawables.useSupportLibrary = true
         missingDimensionStrategy 'react-native-camera', 'general'
+        testBuildType System.getProperty('testBuildType', 'debug')  // This will later be used to control the test apk build type
+        testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
```
Can we add `androidx` dependencies to our RN59 app?

mnzaki replied:
I was surprised by this line as well and intended to investigate it like a week ago, but then totally forgot because, well, it works...... so I guess, to answer your question: yes, looks like it

mnzaki replied:
No actually I take that back. I think this is only for the "instrumentation runner". Detox installs another app separately for running instrumentation. I don't think we can mix in androidx stuff into our own app. (haven't confirmed 100%)

VolkerSchiewe commented on android/build.gradle:
```diff
@@ -3,17 +3,19 @@
         compileSdkVersion = 28
         targetSdkVersion = 28
         supportLibVersion = "28.0.0"
+        kotlinVersion = "1.3.0"
```
Why did you add kotlin?

mnzaki replied:
Requirement by Detox's latest version

mnzaki replied:
hmmm I think I will add some comments to these things

saifahn commented:
## iOS Device

I ran `detox build -c ios.sim.debug && detox test -c ios.sim.debug` and after the dev bundle loaded onto the simulator, it gave the same as above. The 01_landing_spec passed but the others had the device ReferenceError.

There should also be some documentation somewhere regarding the pitfalls - e.g. needing the bundle server started and having the bundle fully loaded before running the tests.

mnzaki commented:
There's also a conflict with the new entropy screen, because the infinite loop animation hangs detox :/

saifahn commented:
I can confirm the hanging on the entropy screen.

Next is:

Input Seed Phrase Screen
      ✓ should show a recoveryMsg (145ms)
      ✓ should show a seedWordFld (1853ms)
      ✓ should show word suggestions based on input prefix (8284ms)
      ✕ should add words to seed phrase on tap (2607ms)
      ✕ should show a restoreAccount button (35ms)
      ✕ should navigate home after a successful restore (10007ms)

'Multiple UI elements matched for given criteria.'

At e2e/utils.ts:42 

```ts
try {
   //console.error('expect with child')
   await expect(el).toBeVisible()
} catch (err) {
```

mnzaki commented:
> There's also a conflict with the new entropy screen, because the infinite loop animation hangs detox :/

I'm using a fork of detox with an unreleased branch to support proper disabling of synchronization, so we can avoid the animation thing.

Another issue with the new entropy screen was the the animation loop would continue even after the animation component was unmounted.

These issues have been addressed in the last couple commits

mnzaki commented:
I have moved the detox fork to the @jolocom organization and updated package.json.
Note that the fork contains a build of the branch https://github.com/jolocom/Detox/tree/jolocom on the tag `detox-v9000.0.1-gitpkg` using [gitpkg](https://github.com/ramasilveyra/gitpkg).

The `jolocom` branch currently only has this PR merged in https://github.com/wix/Detox/pull/1521

mnzaki commented:
> I was under the impression that clearTimeout clears all the memory associated with the callback and so de-allocates the animation stuff, but there's a pretty good chance I'm wrong, and also the flag method is much clearer to understand/read, noice

`clearTimeout` would only remove the timeout from the queue, if it hasn't already been executed. And that timeout was just the one that starts the first pulse. The callback which is passed to `.start` though is used internally and scheduled in a different manner, there are definitely other references to it and wouldn't get garbage collected (and anyway most probably the initial timeout will have executed anyway since the delay is `0` :D)

Exulansis commented:
Tested as well, this looks good, outside of the merge conflicts I think this can be merged.

An extra issue should be created to add documentation on how to run this. I found the comments left by @saifahn really helpful.
